In [1]:
import pandas as pd
import numpy as np

In [2]:
#Load customers orders and payments
#from google.colab import drive
#drive.mount('/content/drive')
customers = pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_customers_dataset.csv')
orders = pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_orders_dataset.csv')
payments = pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_order_payments_dataset.csv')

print("All datasets loaded successfully!")


All datasets loaded successfully!


Check data

In [3]:
print(customers.shape)
print(orders.shape)
print(payments.shape)

(99441, 5)
(99441, 8)
(103886, 5)


In [4]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [5]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,02-10-2017 10:56,02-10-2017 11:07,04-10-2017 19:55,10-10-2017 21:25,18-10-2017 00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,24-07-2018 20:41,26-07-2018 03:24,26-07-2018 14:31,07-08-2018 15:27,13-08-2018 00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,08-08-2018 08:38,08-08-2018 08:55,08-08-2018 13:50,17-08-2018 18:06,04-09-2018 00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,18-11-2017 19:28,18-11-2017 19:45,22-11-2017 13:39,02-12-2017 00:28,15-12-2017 00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,13-02-2018 21:18,13-02-2018 22:20,14-02-2018 19:46,16-02-2018 18:17,26-02-2018 00:00


In [6]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [7]:
customers.isnull().sum()

,0
customer_id,0
customer_unique_id,0
customer_zip_code_prefix,0
customer_city,0
customer_state,0


In [8]:
orders.isnull().sum()

,0
order_id,0
customer_id,0
order_status,0
order_purchase_timestamp,0
order_approved_at,160
order_delivered_carrier_date,1783
order_delivered_customer_date,2965
order_estimated_delivery_date,0


In [9]:
payments.isnull().sum()

,0
order_id,0
payment_sequential,0
payment_type,0
payment_installments,0
payment_value,0


In [10]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [11]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [12]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [13]:
#convert into date time

orders['order_purchase_timestamp'] = pd.to_datetime(
    orders['order_purchase_timestamp'],
    dayfirst=True
)

In [14]:
orders['order_purchase_timestamp']

,order_purchase_timestamp
0,2017-10-02 10:56:00
1,2018-07-24 20:41:00
2,2018-08-08 08:38:00
3,2017-11-18 19:28:00
4,2018-02-13 21:18:00
...,...
99436,2017-03-09 09:54:00
99437,2018-02-06 12:58:00
99438,2017-08-27 14:46:00
99439,2018-01-08 21:28:00


In [15]:
orders['order_status'].value_counts()

,count
order_status,
delivered,96478
shipped,1107
canceled,625
unavailable,609
invoiced,314
processing,301
created,5
approved,2


In [16]:
orders = orders[
    orders['order_status'] == 'delivered'
]

In [17]:
orders['order_status'].value_counts()

,count
order_status,
delivered,96478


In [18]:
orders.shape

(96478, 8)

**# Merge Customers and Orders**


In [19]:
customer_orders = pd.merge(
    customers, orders,
    on = 'customer_id',
    how='inner'
)

In [20]:
customer_orders.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,00e7ee1b050b8499577073aeb2a297a1,delivered,2017-05-16 15:05:00,16-05-2017 15:22,23-05-2017 10:47,25-05-2017 10:35,05-06-2017 00:00
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP,29150127e6685892b6eab3eec79f59c7,delivered,2018-01-12 20:48:00,12-01-2018 20:58,15-01-2018 17:14,29-01-2018 12:41,06-02-2018 00:00
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP,b2059ed67ce144a36e2aa97d2c9e9ad2,delivered,2018-05-19 16:07:00,20-05-2018 16:19,11-06-2018 14:31,14-06-2018 17:58,13-06-2018 00:00
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP,951670f92359f4fe4a63112aa7306eba,delivered,2018-03-13 16:06:00,13-03-2018 17:29,27-03-2018 23:22,28-03-2018 16:04,10-04-2018 00:00
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP,6b7d50bd145f6fc7f33cebabd7e49d0f,delivered,2018-07-29 09:51:00,29-07-2018 10:10,30-07-2018 15:16,09-08-2018 20:55,15-08-2018 00:00


In [21]:
customer_orders.shape

(96478, 12)

In [22]:
list(customer_orders.columns)


['customer_id',
 'customer_unique_id',
 'customer_zip_code_prefix',
 'customer_city',
 'customer_state',
 'order_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date']

In [23]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [24]:
customer_orders_payments =pd.merge(
    customer_orders,
    payments,
    on='order_id',
    how='inner'
)

In [25]:
customer_orders_payments.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,payment_sequential,payment_type,payment_installments,payment_value
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,00e7ee1b050b8499577073aeb2a297a1,delivered,2017-05-16 15:05:00,16-05-2017 15:22,23-05-2017 10:47,25-05-2017 10:35,05-06-2017 00:00,1,credit_card,2,146.87
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP,29150127e6685892b6eab3eec79f59c7,delivered,2018-01-12 20:48:00,12-01-2018 20:58,15-01-2018 17:14,29-01-2018 12:41,06-02-2018 00:00,1,credit_card,8,335.48
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP,b2059ed67ce144a36e2aa97d2c9e9ad2,delivered,2018-05-19 16:07:00,20-05-2018 16:19,11-06-2018 14:31,14-06-2018 17:58,13-06-2018 00:00,1,credit_card,7,157.73
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP,951670f92359f4fe4a63112aa7306eba,delivered,2018-03-13 16:06:00,13-03-2018 17:29,27-03-2018 23:22,28-03-2018 16:04,10-04-2018 00:00,1,credit_card,1,173.30
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP,6b7d50bd145f6fc7f33cebabd7e49d0f,delivered,2018-07-29 09:51:00,29-07-2018 10:10,30-07-2018 15:16,09-08-2018 20:55,15-08-2018 00:00,1,credit_card,8,252.25


In [26]:
customer_orders_payments.shape

(100756, 16)

In [27]:
payment_summary = (
    payments.groupby('order_id', as_index=False)
    ['payment_value'].sum()
)

In [28]:
payment_summary.head()

,order_id,payment_value
0,00010242fe8c5a6d1ba2dd792cb16214,72.19
1,00018f77f2f0320c557190d7a144bdd3,259.83
2,000229ec398224ef6ca0657da4fc703e,216.87
3,00024acbcdf0a6daa1e931b038114c75,25.78
4,00042b26cf59d7ce69dfabb4e55b4fd9,218.04


In [29]:
payment_summary['order_id'].nunique()

99440

In [30]:
len(payment_summary)

99440

In [31]:
payment_summary.shape

(99440, 2)

In [32]:
rfm_base=pd.merge(
    customer_orders,
    payment_summary,
    on='order_id',
    how='inner'
)

In [33]:
rfm_base.shape

(96477, 13)

In [34]:
rfm_base.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,payment_value
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,00e7ee1b050b8499577073aeb2a297a1,delivered,2017-05-16 15:05:00,16-05-2017 15:22,23-05-2017 10:47,25-05-2017 10:35,05-06-2017 00:00,146.87
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP,29150127e6685892b6eab3eec79f59c7,delivered,2018-01-12 20:48:00,12-01-2018 20:58,15-01-2018 17:14,29-01-2018 12:41,06-02-2018 00:00,335.48
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP,b2059ed67ce144a36e2aa97d2c9e9ad2,delivered,2018-05-19 16:07:00,20-05-2018 16:19,11-06-2018 14:31,14-06-2018 17:58,13-06-2018 00:00,157.73
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP,951670f92359f4fe4a63112aa7306eba,delivered,2018-03-13 16:06:00,13-03-2018 17:29,27-03-2018 23:22,28-03-2018 16:04,10-04-2018 00:00,173.30
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP,6b7d50bd145f6fc7f33cebabd7e49d0f,delivered,2018-07-29 09:51:00,29-07-2018 10:10,30-07-2018 15:16,09-08-2018 20:55,15-08-2018 00:00,252.25


In [35]:
rfm_base[
    [
        'customer_unique_id',
        'order_id',
        'order_purchase_timestamp',
        'payment_value'
    ]
].head()

,customer_unique_id,order_id,order_purchase_timestamp,payment_value
0,861eff4711a542e4b93843c6dd7febb0,00e7ee1b050b8499577073aeb2a297a1,2017-05-16 15:05:00,146.87
1,290c77bc529b7ac935b93aa66c333dc3,29150127e6685892b6eab3eec79f59c7,2018-01-12 20:48:00,335.48
2,060e732b5b29e8181a18229c7b0b2b5e,b2059ed67ce144a36e2aa97d2c9e9ad2,2018-05-19 16:07:00,157.73
3,259dac757896d24d7702b9acbbff3f3c,951670f92359f4fe4a63112aa7306eba,2018-03-13 16:06:00,173.30
4,345ecd01c38d18a9036ed96c73b8d066,6b7d50bd145f6fc7f33cebabd7e49d0f,2018-07-29 09:51:00,252.25


In [36]:
#For Recency

snapshot_date =rfm_base['order_purchase_timestamp'].max()
print(snapshot_date)


2018-08-29 15:00:00


In [37]:
rfm =rfm_base.groupby('customer_unique_id').agg({
    'order_purchase_timestamp': 'max',
    'order_id':'nunique',
    'payment_value':'sum'
}).reset_index()

In [38]:
rfm.head()

,customer_unique_id,order_purchase_timestamp,order_id,payment_value
0,0000366f3b9a7992bf8c76cfdf3221e2,2018-05-10 10:56:00,1,141.90
1,0000b849f77a49e4a4ce2b2a4ca5be3f,2018-05-07 11:11:00,1,27.19
2,0000f46a3911fa3c0805444483337064,2017-03-10 21:05:00,1,86.22
3,0000f6ccb0745a6a4b88665a16c9f078,2017-10-12 20:29:00,1,43.62
4,0004aac84e0df4da2b147fca70cf8255,2017-11-14 19:45:00,1,196.89


In [39]:
rfm.columns = [
    'customer_unique_id',
    'last_purchase_date',
    'frequency',
    'monetary'
]

In [40]:
rfm.head()

,customer_unique_id,last_purchase_date,frequency,monetary
0,0000366f3b9a7992bf8c76cfdf3221e2,2018-05-10 10:56:00,1,141.90
1,0000b849f77a49e4a4ce2b2a4ca5be3f,2018-05-07 11:11:00,1,27.19
2,0000f46a3911fa3c0805444483337064,2017-03-10 21:05:00,1,86.22
3,0000f6ccb0745a6a4b88665a16c9f078,2017-10-12 20:29:00,1,43.62
4,0004aac84e0df4da2b147fca70cf8255,2017-11-14 19:45:00,1,196.89


In [41]:
rfm.shape

(93357, 4)

In [42]:
rfm['recency']=(
    snapshot_date-rfm['last_purchase_date']

).dt.days

In [43]:
rfm.head()

,customer_unique_id,last_purchase_date,frequency,monetary,recency
0,0000366f3b9a7992bf8c76cfdf3221e2,2018-05-10 10:56:00,1,141.90,111
1,0000b849f77a49e4a4ce2b2a4ca5be3f,2018-05-07 11:11:00,1,27.19,114
2,0000f46a3911fa3c0805444483337064,2017-03-10 21:05:00,1,86.22,536
3,0000f6ccb0745a6a4b88665a16c9f078,2017-10-12 20:29:00,1,43.62,320
4,0004aac84e0df4da2b147fca70cf8255,2017-11-14 19:45:00,1,196.89,287


In [44]:
rfm[['recency','frequency','monetary']].describe()

,recency,frequency,monetary
count,93357.000000,93357.000000,93357.000000
mean,236.937155,1.033420,165.198772
std,152.584344,0.209099,226.314579
min,0.000000,1.000000,9.590000
25%,113.000000,1.000000,63.060000
50%,218.000000,1.000000,107.780000
75%,345.000000,1.000000,182.560000
max,694.000000,15.000000,13664.080000


In [45]:
rfm['R_score'] = pd.qcut(
    rfm['recency'],
    q=5,
    labels=[5,4,3,2,1]
)

In [46]:
rfm['R_score'].value_counts()

,count
R_score,
5,18728
3,18709
4,18705
1,18638
2,18577


In [47]:
rfm['frequency'].value_counts().head(20)

,count
frequency,
1,90556
2,2573
3,181
4,28
5,9
6,5
7,3
9,1
15,1


In [48]:
def frequency_score(x):
    if x >= 3:
        return 5
    elif x == 2:
        return 4
    else:
        return 1

rfm['F_score'] = rfm['frequency'].apply(frequency_score)

In [49]:
rfm['F_score'].value_counts()

,count
F_score,
1,90556
4,2573
5,228


In [50]:
rfm['frequency'].describe()

,frequency
count,93357.000000
mean,1.033420
std,0.209099
min,1.000000
25%,1.000000
50%,1.000000
75%,1.000000
max,15.000000


In [51]:
#rename
def frequency_score(x):
    if x == 1:
        return "One-Time Buyer"
    elif x == 2:
        return "Returning Customer"
    else:
        return "Loyal Customer"

rfm['frequency_segment'] = rfm['frequency'].apply(frequency_score)

In [52]:
rfm['frequency_segment'].value_counts()

,count
frequency_segment,
One-Time Buyer,90556
Returning Customer,2573
Loyal Customer,228


In [53]:
rfm['M_score'] = pd.qcut(
    rfm['monetary'],
    q=5,
    labels=[1,2,3,4,5]
)

In [54]:
rfm['M_score'].value_counts()

,count
M_score,
2,18683
4,18675
1,18673
5,18668
3,18658


In [55]:
rfm['M_score'].value_counts()

,count
M_score,
2,18683
4,18675
1,18673
5,18668
3,18658


In [56]:
#for frequency
rfm['F_score'] = np.where(
    rfm['frequency'] == 1,
    1,
    5
)

In [57]:
rfm['frequency_segment'].value_counts()

,count
frequency_segment,
One-Time Buyer,90556
Returning Customer,2573
Loyal Customer,228


In [58]:
#check
(rfm['frequency'] > 1).sum()

np.int64(2801)

In [59]:
# Repeat Purchase Rate

((rfm['frequency'] > 1).sum() / len(rfm) * 100)

np.float64(3.0003106355174225)

In [60]:
rfm['RM_score'] = (
    rfm['R_score'].astype(str) +
    rfm['M_score'].astype(str)
)

In [61]:
rfm[['recency','monetary']].describe()

,recency,monetary
count,93357.000000,93357.000000
mean,236.937155,165.198772
std,152.584344,226.314579
min,0.000000,9.590000
25%,113.000000,63.060000
50%,218.000000,107.780000
75%,345.000000,182.560000
max,694.000000,13664.080000


In [62]:
#Combined View
rfm[['R_score','M_score']].head()

,R_score,M_score
0,4,4
1,4,1
2,1,2
3,2,1
4,2,4


In [63]:
rfm.groupby('frequency_segment').agg({
    'monetary':'mean',
    'recency':'mean'
}).round(2)

,monetary,recency
frequency_segment,,
Loyal Customer,506.79,201.28
One-Time Buyer,160.76,237.48
Returning Customer,291.03,220.89


In [64]:
# Best Customers
rfm[
    (rfm['R_score'] == 5) &
    (rfm['M_score'] == 5)
].shape

(3815, 10)

In [65]:
#Worst Customers
rfm[
    (rfm['R_score'] == 1) &
    (rfm['M_score'] == 1)
].shape

(3888, 10)

In [66]:
#heatmap matrix
rfm.groupby(['R_score','M_score']).size().unstack(fill_value=0)

/tmp/ipykernel_1986/2223196038.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  rfm.groupby(['R_score','M_score']).size().unstack(fill_value=0)


M_score,1,2,3,4,5
R_score,,,,,
5,3714,3543,3765,3891,3815
4,3618,3545,3812,3904,3826
3,3905,3652,3733,3857,3562
2,3548,4001,3653,3621,3754
1,3888,3942,3695,3402,3711


In [67]:
#final customer segment column.
def segment_customer(row):

    if row['R_score'] >= 4 and row['M_score'] >= 4:
        return 'Champion'

    elif row['R_score'] >= 4 and row['M_score'] <= 3:
        return 'Potential Loyalist'

    elif row['R_score'] <= 2 and row['M_score'] >= 4:
        return 'At Risk'

    elif row['R_score'] <= 2 and row['M_score'] <= 2:
        return 'Lost Customer'

    else:
        return 'Regular Customer'

In [68]:
rfm['customer_segment'] = rfm.apply(
    segment_customer,
    axis=1
)

In [69]:
rfm['customer_segment'].value_counts()

,count
customer_segment,
Regular Customer,26057
Potential Loyalist,21997
Champion,15436
Lost Customer,15379
At Risk,14488


In [70]:
pd.crosstab(
    rfm['customer_segment'],
    rfm['frequency_segment']
)

frequency_segment,Loyal Customer,One-Time Buyer,Returning Customer
customer_segment,,,
At Risk,60,13739,689
Champion,117,14446,873
Lost Customer,3,15298,78
Potential Loyalist,5,21778,214
Regular Customer,43,25295,719


In [71]:
rfm.to_csv('rfm_customer_segments.csv', index=False)

# Visualization

In [72]:
import plotly.express as px

In [73]:
#1. Customer Value Segmentation
segment_counts = (
    rfm['customer_segment']
    .value_counts()
    .reset_index()
)

segment_counts.columns = ['Segment', 'Customers']

fig = px.bar(
    segment_counts,
    x='Customers',
    y='Segment',
    orientation='h',
    color='Segment',
    title='Customer Value Segmentation',
    color_discrete_sequence=px.colors.qualitative.Bold
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis_title='Number of Customers',
    yaxis_title='',
    showlegend=False
)

fig.show()

In [74]:
# 2. RFM Heatmap


heatmap = (
    rfm.groupby(['R_score', 'M_score'])
       .size()
       .reset_index(name='Customers')
)

fig = px.density_heatmap(
    heatmap,
    x='M_score',
    y='R_score',
    z='Customers',
    text_auto=True,
    color_continuous_scale='RdYlGn',
    title='RFM Customer Distribution Heatmap'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis_title='Monetary Score',
    yaxis_title='Recency Score'
)

fig.show()

/tmp/ipykernel_1986/1453538711.py:5: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [75]:
# 3: Average Spend by Customer Segment
avg_spend = (
    rfm.groupby('customer_segment')['monetary']
       .mean()
       .sort_values(ascending=False)
       .reset_index()
)

fig = px.bar(
    avg_spend,
    x='customer_segment',
    y='monetary',
    color='customer_segment',
    title='Average Spend by Customer Segment',
    color_discrete_sequence=px.colors.qualitative.Set2
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis_title='Customer Segment',
    yaxis_title='Average Spend',
    showlegend=False
)

fig.show()

In [76]:
#4: Frequency Segment Distribution
retention_counts = (
    rfm['frequency_segment']
    .value_counts()
    .reset_index()
)

retention_counts.columns = ['Segment', 'Customers']

fig = px.bar(
    retention_counts,
    x='Segment',
    y='Customers',
    color='Segment',
    title='Frequency Segment Distribution',
    color_discrete_map={
        'One-Time Buyer': '#E63946',
        'Returning Customer': '#457B9D',
        'Loyal Customer': '#2A9D8F'
    }
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    yaxis_title='Number of Customers',
    showlegend=False
)

fig.show()

In [77]:
#5. Customer Segment vs Retention Segment
cross_tab = pd.crosstab(
    rfm['customer_segment'],
    rfm['frequency_segment']
)

fig = px.imshow(
    cross_tab,
    text_auto=True,
    color_continuous_scale='Blues',
    aspect='auto',
    title='Customer Value Segment vs Retention Segment'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)

fig.show()

# **Product Analytics**


In [78]:
# Load products, order items and category translation datasets


products =  pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_products_dataset.csv')
order_items = pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_order_items_dataset.csv')
translation = pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/product_category_name_translation.csv')

print("All datasets loaded successfully!")

All datasets loaded successfully!


In [79]:
# Check dataset dimensions

print("Products:", products.shape)
print("Order Items:", order_items.shape)
print("Translation:", translation.shape)

Products: (32951, 9)
Order Items: (112650, 7)
Translation: (71, 2)


In [80]:
# Display first 5 rows

products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [81]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,19-09-2017 09:45,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,03-05-2017 11:05,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,18-01-2018 14:48,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,15-08-2018 10:10,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,13-02-2017 13:57,199.90,18.14


In [82]:
translation.head()


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [83]:
# Check columns

print(products.columns.tolist())
print(order_items.columns.tolist())
print(translation.columns.tolist())

['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
['product_category_name', 'product_category_name_english']


In [84]:
# Missing values in products

products.isnull().sum()

,0
product_id,0
product_category_name,610
product_name_lenght,610
product_description_lenght,610
product_photos_qty,610
product_weight_g,2
product_length_cm,2
product_height_cm,2
product_width_cm,2


In [85]:
# Missing values in order_items

order_items.isnull().sum()

,0
order_id,0
order_item_id,0
product_id,0
seller_id,0
shipping_limit_date,0
price,0
freight_value,0


In [86]:
# Missing values in translation

translation.isnull().sum()

,0
product_category_name,0
product_category_name_english,0


In [87]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [88]:
translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [89]:
products_en=products.merge(
    translation,
    on='product_category_name',
    how='left'
)

In [90]:
products_en.shape

(32951, 10)

In [91]:
products_en.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0,perfumery
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0,art
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0,sports_leisure
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0,baby
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0,housewares


In [92]:
products_en.columns.tolist()

['product_id',
 'product_category_name',
 'product_name_lenght',
 'product_description_lenght',
 'product_photos_qty',
 'product_weight_g',
 'product_length_cm',
 'product_height_cm',
 'product_width_cm',
 'product_category_name_english']

In [93]:
#Product Sales Table

product_sales = order_items.merge(
    products_en,
    on='product_id',
    how='left'
)

product_sales.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,19-09-2017 09:45,58.90,13.29,cool_stuff,58.0,598.0,4.0,650.0,28.0,9.0,14.0,cool_stuff
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,03-05-2017 11:05,239.90,19.93,pet_shop,56.0,239.0,2.0,30000.0,50.0,30.0,40.0,pet_shop
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,18-01-2018 14:48,199.00,17.87,moveis_decoracao,59.0,695.0,2.0,3050.0,33.0,13.0,33.0,furniture_decor
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,15-08-2018 10:10,12.99,12.79,perfumaria,42.0,480.0,1.0,200.0,16.0,10.0,15.0,perfumery
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,13-02-2017 13:57,199.90,18.14,ferramentas_jardim,59.0,409.0,1.0,3750.0,35.0,40.0,30.0,garden_tools


In [94]:
product_sales.shape

(112650, 16)

In [95]:
product_sales.columns.tolist()

['order_id',
 'order_item_id',
 'product_id',
 'seller_id',
 'shipping_limit_date',
 'price',
 'freight_value',
 'product_category_name',
 'product_name_lenght',
 'product_description_lenght',
 'product_photos_qty',
 'product_weight_g',
 'product_length_cm',
 'product_height_cm',
 'product_width_cm',
 'product_category_name_english']

In [96]:
product_sales.shape



(112650, 16)

In [97]:
#null values
#1,627 order items do not have a mapped product category.

product_sales['product_category_name_english'].isnull().sum()

np.int64(1627)

In [98]:
(1627 / len(product_sales)) * 100

1.4442964935641367

In [99]:
product_sales_clean = product_sales.dropna(
    subset=['product_category_name_english']
)

In [100]:
product_sales_clean.shape

(111023, 16)

In [101]:
product_sales_clean['product_category_name_english'].isnull().sum()

np.int64(0)

**Pareto Analysis**

In [102]:

pareto = (
    product_sales_clean
    .groupby('product_category_name_english')['price']
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)
pareto.columns = [ 'category','revenue']

In [103]:
#Revenue Contribution %
pareto['revenue_pct'] = (
    pareto['revenue']
    / pareto['revenue'].sum()
) * 100
pareto.head()

,category,revenue,revenue_pct
0,health_beauty,1258681.34,9.388524
1,watches_gifts,1205005.68,8.988157
2,bed_bath_table,1036988.68,7.734915
3,sports_leisure,988048.97,7.369873
4,computers_accessories,911954.32,6.802282


In [104]:
#Cumulative Revenue %

pareto['cumulative_pct'] = (
    pareto['revenue_pct']
    .cumsum()
)
pareto.head(15)

,category,revenue,revenue_pct,cumulative_pct
0,health_beauty,1258681.34,9.388524,9.388524
1,watches_gifts,1205005.68,8.988157,18.376681
2,bed_bath_table,1036988.68,7.734915,26.111596
3,sports_leisure,988048.97,7.369873,33.481470
4,computers_accessories,911954.32,6.802282,40.283752
5,furniture_decor,729762.49,5.443310,45.727062
6,cool_stuff,635290.85,4.738645,50.465706
7,housewares,632248.66,4.715953,55.181659
8,auto,592720.11,4.421109,59.602768
9,garden_tools,485256.46,3.619536,63.222304


In [105]:
print("Total Categories:", len(pareto))

print(
    "Categories contributing to first 80% revenue:",
    len(pareto[pareto['cumulative_pct'] <= 80])
)


Total Categories: 71
Categories contributing to first 80% revenue: 16


In [106]:
(16 / 71) * 100

22.535211267605636

In [107]:
import plotly.graph_objects as go

pareto_top = pareto.head(20)

fig = go.Figure()

# Revenue Contribution Bars
fig.add_bar(
    x=pareto_top['category'],
    y=pareto_top['revenue_pct'],
    name='Revenue Contribution (%)'
)

# Cumulative Revenue Line
fig.add_scatter(
    x=pareto_top['category'],
    y=pareto_top['cumulative_pct'],
    mode='lines+markers',
    name='Cumulative Revenue (%)',
    yaxis='y2'
)

# 80% Reference Line
fig.add_hline(
    y=80,
    line_dash='dash',
    line_color='red',
    annotation_text='80% Threshold'
)

fig.update_layout(
    title='Pareto Analysis: Revenue Concentration by Product Category',
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',

    xaxis=dict(
        title='Category',
        tickangle=-45,
        showgrid=False,
        showline=True,
        linecolor='black'
    ),

    yaxis=dict(
        title='Revenue Contribution (%)',
        showgrid=True,
        gridcolor='#E5E5E5',
        showline=True,
        linecolor='black'
    ),

    yaxis2=dict(
        title='Cumulative Revenue (%)',
        overlaying='y',
        side='right',
        range=[0,100]
    ),

    legend=dict(
        orientation='h',
        y=1.08
    )
)

fig.show()

Although the marketplace offers 71 product categories, only 16 categories account for approximately 80% of total revenue. This indicates that revenue generation is highly concentrated within a relatively small portion of the product portfolio. Strategic focus on these high-performing categories can significantly influence overall business performance, while lower-contributing categories may represent diversification opportunities or portfolio optimization candidates.

**Product Portfolio Matrix**

What role does each category play in the product portfolio?

In [108]:
portfolio = (
    product_sales_clean
    .groupby('product_category_name_english')
    .agg(
        revenue=('price', 'sum'),
        items_sold=('product_id', 'count'),
        avg_price=('price', 'mean')
    )
    .reset_index()
)

portfolio.head()

,product_category_name_english,revenue,items_sold,avg_price
0,agro_industry_and_commerce,72530.47,212,342.124858
1,air_conditioning,55024.96,297,185.269226
2,art,24202.64,209,115.802105
3,arts_and_craftmanship,1814.01,24,75.583750
4,audio,50688.50,364,139.254121


In [109]:
portfolio.describe()

,revenue,items_sold,avg_price
count,7.100000e+01,71.000000,71.000000
mean,1.888253e+05,1563.704225,144.778802
std,3.037052e+05,2606.545647,147.546088
min,2.832900e+02,2.000000,25.342333
25%,9.171185e+03,93.000000,79.566250
50%,4.685688e+04,281.000000,114.344285
75%,2.029071e+05,1819.000000,147.402966
max,1.258681e+06,11115.000000,1098.340542


In [110]:
portfolio.sort_values(
    'revenue',
    ascending=False
).head(10)

,product_category_name_english,revenue,items_sold,avg_price
43,health_beauty,1258681.34,9670,130.163531
70,watches_gifts,1205005.68,5991,201.135984
7,bed_bath_table,1036988.68,11115,93.296327
65,sports_leisure,988048.97,8641,114.344285
15,computers_accessories,911954.32,7827,116.513903
39,furniture_decor,729762.49,8334,87.564494
20,cool_stuff,635290.85,3796,167.357969
49,housewares,632248.66,6964,90.788148
5,auto,592720.11,4235,139.957523
42,garden_tools,485256.46,4347,111.630196


In [111]:
portfolio.sort_values(
    'avg_price',
    ascending=False
).head(10)

,product_category_name_english,revenue,items_sold,avg_price
14,computers,222963.13,203,1098.340542
64,small_appliances_home_oven_and_coffee,47445.71,76,624.285658
45,home_appliances_2,113317.74,238,476.124958
0,agro_industry_and_commerce,72530.47,212,342.124858
56,musical_instruments,191498.88,680,281.616000
63,small_appliances,190648.58,679,280.778468
34,fixed_telephony,59583.00,264,225.693182
19,construction_tools_safety,40544.52,194,208.992371
70,watches_gifts,1205005.68,5991,201.135984
1,air_conditioning,55024.96,297,185.269226


In [112]:
price_cutoff = portfolio['avg_price'].median()
volume_cutoff = portfolio['items_sold'].median()

print("Median Price:", round(price_cutoff,2))
print("Median Volume:", round(volume_cutoff,2))

Median Price: 114.34
Median Volume: 281.0


In [113]:
def portfolio_segment(row):

    if row['avg_price'] >= price_cutoff and row['items_sold'] >= volume_cutoff:
        return 'Portfolio Star'

    elif row['avg_price'] < price_cutoff and row['items_sold'] >= volume_cutoff:
        return 'Traffic Driver'

    elif row['avg_price'] >= price_cutoff and row['items_sold'] < volume_cutoff:
        return 'Premium Niche'

    else:
        return 'Low Impact'


portfolio['portfolio_segment'] = (
    portfolio.apply(
        portfolio_segment,
        axis=1
    )
)

portfolio.head()

,product_category_name_english,revenue,items_sold,avg_price,portfolio_segment
0,agro_industry_and_commerce,72530.47,212,342.124858,Premium Niche
1,air_conditioning,55024.96,297,185.269226,Portfolio Star
2,art,24202.64,209,115.802105,Premium Niche
3,arts_and_craftmanship,1814.01,24,75.583750,Low Impact
4,audio,50688.50,364,139.254121,Portfolio Star


In [114]:
portfolio['portfolio_segment'].value_counts()

,count
portfolio_segment,
Portfolio Star,22
Low Impact,21
Premium Niche,14
Traffic Driver,14


In [115]:
#Portfolio Stars
portfolio[
    portfolio['portfolio_segment'] == 'Portfolio Star'
].sort_values(
    'revenue',
    ascending=False
).head(10)

,product_category_name_english,revenue,items_sold,avg_price,portfolio_segment
43,health_beauty,1258681.34,9670,130.163531,Portfolio Star
70,watches_gifts,1205005.68,5991,201.135984,Portfolio Star
65,sports_leisure,988048.97,8641,114.344285,Portfolio Star
15,computers_accessories,911954.32,7827,116.513903,Portfolio Star
20,cool_stuff,635290.85,3796,167.357969,Portfolio Star
5,auto,592720.11,4235,139.957523,Portfolio Star
69,toys,483946.60,4117,117.548360,Portfolio Star
6,baby,411764.89,3065,134.344173,Portfolio Star
59,perfumery,399124.87,3419,116.737312,Portfolio Star
57,office_furniture,273960.70,1691,162.011059,Portfolio Star


In [116]:
#Traffic Drivers
portfolio[
    portfolio['portfolio_segment'] == 'Traffic Driver'
].sort_values(
    'items_sold',
    ascending=False
).head(10)

,product_category_name_english,revenue,items_sold,avg_price,portfolio_segment
7,bed_bath_table,1036988.68,11115,93.296327,Traffic Driver
39,furniture_decor,729762.49,8334,87.564494,Traffic Driver
49,housewares,632248.66,6964,90.788148,Traffic Driver
68,telephony,323667.53,4545,71.213978,Traffic Driver
42,garden_tools,485256.46,4347,111.630196,Traffic Driver
26,electronics,160246.74,2767,57.913531,Traffic Driver
66,stationery,230943.23,2517,91.753369,Traffic Driver
28,fashion_bags_accessories,152823.54,2031,75.245465,Traffic Driver
60,pet_shop,214315.41,1947,110.074684,Traffic Driver
44,home_appliances,80171.53,771,103.983826,Traffic Driver


In [117]:
#Premium Niches
portfolio[
    portfolio['portfolio_segment'] == 'Premium Niche'
].sort_values(
    'avg_price',
    ascending=False
).head(10)

,product_category_name_english,revenue,items_sold,avg_price,portfolio_segment
14,computers,222963.13,203,1098.340542,Premium Niche
64,small_appliances_home_oven_and_coffee,47445.71,76,624.285658,Premium Niche
45,home_appliances_2,113317.74,238,476.124958,Premium Niche
0,agro_industry_and_commerce,72530.47,212,342.124858,Premium Niche
34,fixed_telephony,59583.00,264,225.693182,Premium Niche
19,construction_tools_safety,40544.52,194,208.992371,Premium Niche
38,furniture_bedroom,20028.78,109,183.750275,Premium Niche
55,music,6034.35,38,158.798684,Premium Niche
22,costruction_tools_tools,15903.95,103,154.407282,Premium Niche
50,industry_commerce_and_business,39669.61,268,148.020933,Premium Niche


# **Portfolio Segment Chart**

In [118]:
import numpy as np

portfolio['revenue_scaled'] = np.sqrt(portfolio['revenue'])
fig = px.scatter(
    portfolio,
    x='avg_price',
    y='items_sold',
    size='revenue_scaled',
    color='portfolio_segment',
    hover_name='product_category_name_english',
    size_max=40,
    title='Product Portfolio Matrix'
)
fig.update_xaxes(type='log')
fig.update_yaxes(type='log')

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)


price_cutoff = portfolio['avg_price'].median()
volume_cutoff = portfolio['items_sold'].median()

fig.add_vline(
    x=price_cutoff,
    line_dash='dash'
)

fig.add_hline(
    y=volume_cutoff,
    line_dash='dash'
)

fig.show()


# **Freight Efficiency Analysis**

In [119]:
freight_efficiency = (
    product_sales_clean
    .groupby('product_category_name_english')
    .agg(
        avg_price=('price', 'mean'),
        avg_freight=('freight_value', 'mean'),
        items_sold=('product_id', 'count')
    )
    .reset_index()
)

In [120]:
freight_efficiency['freight_ratio'] = (
    freight_efficiency['avg_freight']
    / freight_efficiency['avg_price']
) * 100

In [121]:
freight_efficiency = freight_efficiency[
    freight_efficiency['items_sold'] >= 50
]

In [122]:
freight_efficiency.sort_values(
    'freight_ratio',
    ascending=False
).head(15)

,product_category_name_english,avg_price,avg_freight,items_sold,freight_ratio
12,christmas_supplies,57.521699,21.106536,153,36.693172
62,signaling_and_security,108.086583,32.702613,199,30.255941
37,food_drink,54.602446,16.215791,278,29.697921
26,electronics,57.913531,16.833509,2767,29.066626
40,furniture_living_room,137.011054,35.722008,503,26.072355
51,kitchen_dining_laundry_garden_furniture,164.869644,42.702598,281,25.900825
24,drinks,59.178628,15.148417,379,25.597783
57,office_furniture,162.011059,40.551124,1691,25.029849
36,food,57.634137,14.256922,510,24.736939
39,furniture_decor,87.564494,20.728258,8334,23.671989


In [123]:
top_freight = (
    freight_efficiency
    .sort_values(
        'freight_ratio',
        ascending=False
    )
    .head(10)
)

In [124]:
import plotly.express as px

fig = px.bar(
    top_freight.sort_values('freight_ratio'),
    x='freight_ratio',
    y='product_category_name_english',
    orientation='h',
    color='freight_ratio',
    title='Top Categories by Freight-to-Price Ratio (%)',
    color_continuous_scale='Reds'
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    coloraxis_showscale=False,
    xaxis_title='Freight Ratio (%)',
    yaxis_title='Category'
)

fig.show()

Several high-volume categories, including Electronics, Furniture Decor, Housewares, and Telephony, exhibit freight ratios above 20%, indicating that logistics costs represent a substantial portion of product value. While these categories contribute meaningful sales volume, improving shipping efficiency could have a direct impact on profitability.

In [125]:
order_items.groupby('order_id')['product_id'].count().describe()

,product_id
count,98666.000000
mean,1.141731
std,0.538452
min,1.000000
25%,1.000000
50%,1.000000
75%,1.000000
max,21.000000


# **Market Basket Analysis**

In [126]:
order_items.groupby('order_id')['product_id'].count().describe()

,product_id
count,98666.000000
mean,1.141731
std,0.538452
min,1.000000
25%,1.000000
50%,1.000000
75%,1.000000
max,21.000000


In [127]:
#check for multiple products
(order_items.groupby('order_id')['product_id'].count() > 1).sum()

np.int64(9803)

In [128]:
(
(order_items.groupby('order_id')['product_id'].count() > 1).sum()
/
order_items['order_id'].nunique()
) * 100

np.float64(9.93554010500071)

In [129]:
basket_data = (
    product_sales_clean
    .groupby('order_id')
    ['product_category_name_english']
    .apply(list)
    .reset_index()
)

basket_data.head()

,order_id,product_category_name_english
0,00010242fe8c5a6d1ba2dd792cb16214,[cool_stuff]
1,00018f77f2f0320c557190d7a144bdd3,[pet_shop]
2,000229ec398224ef6ca0657da4fc703e,[furniture_decor]
3,00024acbcdf0a6daa1e931b038114c75,[perfumery]
4,00042b26cf59d7ce69dfabb4e55b4fd9,[garden_tools]


In [130]:
basket_data[
    basket_data['product_category_name_english'].apply(len) > 1
].head(10)

,order_id,product_category_name_english
13,0008288aa423d2a3f00fcb17cd7d8719,"[garden_tools, garden_tools]"
31,00143d0f86d6fbd9f9b38ab440ac16f5,"[sports_leisure, sports_leisure, sports_leisure]"
39,001ab0a7578dd66cd4b0a71f5b6e1e41,"[electronics, electronics, electronics]"
43,001d8f0e34a38c37f7dba2a37d4eba8b,"[health_beauty, health_beauty]"
69,002c9def9c9b951b1bec6d50753c9891,"[housewares, housewares]"
73,002f98c0f7efd42638ed6100ca699b42,"[consoles_games, toys]"
79,003324c70b19a16798817b2b3640e721,"[luggage_accessories, luggage_accessories]"
82,00337fe25a3780b3424d9ad7c5a4b35e,"[bed_bath_table, bed_bath_table]"
91,003822434f91204da0a51fe4cf2aba18,"[perfumery, perfumery]"
100,003f201cdd39cdd59b6447cff2195456,"[computers_accessories, computers_accessories]"


In [131]:
#remove duplicates
basket_data = (
    product_sales_clean
    .groupby('order_id')
    ['product_category_name_english']
    .apply(lambda x: list(set(x)))
    .reset_index()
)

basket_data.head()

,order_id,product_category_name_english
0,00010242fe8c5a6d1ba2dd792cb16214,[cool_stuff]
1,00018f77f2f0320c557190d7a144bdd3,[pet_shop]
2,000229ec398224ef6ca0657da4fc703e,[furniture_decor]
3,00024acbcdf0a6daa1e931b038114c75,[perfumery]
4,00042b26cf59d7ce69dfabb4e55b4fd9,[garden_tools]


In [132]:
basket_data[
    basket_data['product_category_name_english'].apply(len) > 1
].head(20)

,order_id,product_category_name_english
73,002f98c0f7efd42638ed6100ca699b42,"[consoles_games, toys]"
130,005d9a5423d47281ac463a968b3936fb,"[toys, baby]"
474,014405982914c2cde2796ddcf0b8703d,"[sports_leisure, perfumery]"
625,01b1a7fdae9ad1837d6ab861705a1fa5,"[bed_bath_table, housewares]"
675,01cce1175ac3c4a450e3a0f856d02734,"[stationery, garden_tools]"
934,0282e83acfe3e7aaf60c871a4a9aca61,"[market_place, health_beauty]"
1026,02beac5428ccbd30f4b3fbded058289f,"[stationery, computers_accessories]"
1127,0307d42d9c781d32f86cdf1fa5c8670b,"[cool_stuff, sports_leisure]"
1278,0362e923f805ae4dce58fd78c86c96c4,"[kitchen_dining_laundry_garden_furniture, hous..."
1583,0420dbc50fc554e303e4b2d6b39063f6,"[cool_stuff, baby]"


In [133]:
basket_data['product_category_name_english'].apply(len).value_counts()

,count
product_category_name_english,
1,96530
2,711
3,15


After removing duplicate categories within orders, over 99% of orders contained products from a single category. Because cross-category baskets were extremely rare, association analysis produced limited business value and was not pursued further.

Customer purchasing behavior is highly category-focused. More than 99% of orders contain products from a single category, suggesting that customers typically visit the marketplace with a specific purchase intent rather than engaging in broad cross-category shopping.

# **Seller Performance Analytics**

In [134]:
sellers =pd.read_csv('/content/drive/MyDrive/Ecommerce analysis/olist_sellers_dataset.csv')
print('Seller data loaded successfully')

Seller data loaded successfully


In [135]:
sellers.shape

(3095, 4)

In [136]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [137]:
sellers.describe()

,seller_zip_code_prefix
count,3095.000000
mean,32291.059451
std,32713.453830
min,1001.000000
25%,7093.500000
50%,14940.000000
75%,64552.500000
max,99730.000000


In [138]:
sellers.isnull().sum()

,0
seller_id,0
seller_zip_code_prefix,0
seller_city,0
seller_state,0


In [139]:
sellers.columns.tolist()

['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

# **Geographic Seller Analysis**

In [140]:
seller_sales = order_items.merge(
    sellers,
    on='seller_id',
    how='left'
)

seller_sales.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,seller_zip_code_prefix,seller_city,seller_state
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,19-09-2017 09:45,58.90,13.29,27277,volta redonda,SP
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,03-05-2017 11:05,239.90,19.93,3471,sao paulo,SP
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,18-01-2018 14:48,199.00,17.87,37564,borda da mata,MG
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,15-08-2018 10:10,12.99,12.79,14403,franca,SP
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,13-02-2017 13:57,199.90,18.14,87900,loanda,PR


In [141]:
seller_geo = (
    seller_sales
    .groupby('seller_state')
    .agg(
        sellers=('seller_id', 'nunique'),
        revenue=('price', 'sum'),
        orders=('order_id', 'nunique')
    )
    .reset_index()
)

seller_geo['revenue_pct'] = (
    seller_geo['revenue']
    / seller_geo['revenue'].sum()
) * 100

seller_geo = seller_geo.sort_values(
    'revenue',
    ascending=False
)

seller_geo


,seller_state,sellers,revenue,orders,revenue_pct
22,SP,1849,8753396.21,70188,64.402779
15,PR,349,1261887.21,7673,9.284287
8,MG,244,1011564.74,7930,7.442549
16,RJ,171,843984.22,4353,6.209582
20,SC,190,632426.07,3667,4.653051
19,RS,129,378559.54,1989,2.785237
2,BA,19,285561.56,569,2.101008
4,DF,30,97749.48,824,0.719188
13,PE,9,91493.85,406,0.673163
6,GO,40,66399.21,463,0.488530


In [142]:
seller_geo['revenue_per_seller'] = (
    seller_geo['revenue']
    / seller_geo['sellers']
)
seller_geo.head(10)

,seller_state,sellers,revenue,orders,revenue_pct,revenue_per_seller
22,SP,1849,8753396.21,70188,64.402779,4734.124505
15,PR,349,1261887.21,7673,9.284287,3615.722665
8,MG,244,1011564.74,7930,7.442549,4145.757131
16,RJ,171,843984.22,4353,6.209582,4935.580234
20,SC,190,632426.07,3667,4.653051,3328.558263
19,RS,129,378559.54,1989,2.785237,2934.570078
2,BA,19,285561.56,569,2.101008,15029.555789
4,DF,30,97749.48,824,0.719188,3258.316000
13,PE,9,91493.85,406,0.673163,10165.983333
6,GO,40,66399.21,463,0.488530,1659.980250


# **Revenue by Seller State**

In [143]:
import plotly.express as px

fig = px.bar(
    seller_geo.head(10).sort_values('revenue'),
    x='revenue',
    y='seller_state',
    orientation='h',
    title='Revenue Contribution by Seller State',
    color='revenue',
    color_continuous_scale='Blues'
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    coloraxis_showscale=False,
    xaxis_title='Revenue',
    yaxis_title='State'
)

fig.show()

SP dominates marketplace supply,
contributing over 64% of total revenue.

# **Revenue per Seller by State**

In [144]:

seller_geo_filtered = seller_geo[
    seller_geo['sellers'] >= 10
]
fig = px.bar(
         seller_geo_filtered.sort_values(
        'revenue_per_seller',

    )
    .head(10),
    x='revenue_per_seller',
    y='seller_state',
    orientation='h',
    title='Revenue per Seller by State',
    color='revenue_per_seller',
    color_continuous_scale='Oranges'
)

fig.update_traces(
    marker_line_color='black',
    marker_line_width=1
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    coloraxis_showscale=False,
    xaxis_title='Revenue per Seller',
    yaxis_title='State'
)

fig.show()

Rio de Janeiro's sellers generate more revenue per seller than São Paulo, highlighting regional differences in seller productivity and efficiency.

# **Marketplace Dependency Risk**

In [145]:
seller_revenue = (
    seller_sales
    .groupby('seller_id')
    .agg(
        revenue=('price', 'sum'),
        orders=('order_id', 'nunique'),
        items_sold=('product_id', 'count')
    )
    .reset_index()
)

seller_revenue.head()

,seller_id,revenue,orders,items_sold
0,0015a82c2db000af6aaaf3ae2ecb0532,2685.00,3,3
1,001cca7ae9ae17fb1caed9dfb1094831,25080.03,200,239
2,001e6ad469a905060d959994f1b41e4f,250.00,1,1
3,002100f778ceb8431b7a1020ff7ab48f,1234.50,51,55
4,003554e2dce176b5555353e4f3555ac8,120.00,1,1


In [146]:
seller_risk = (
    seller_revenue
    .sort_values(
        'revenue',
        ascending=False
    )
    .reset_index(drop=True)
)

seller_risk['revenue_pct'] = (
    seller_risk['revenue']
    / seller_risk['revenue'].sum()
) * 100

seller_risk['cumulative_pct'] = (
    seller_risk['revenue_pct']
    .cumsum()
)
seller_risk

,seller_id,revenue,orders,items_sold,revenue_pct,cumulative_pct
0,4869f7a5dfa277a7dca6462dcf3b52b2,229472.63,1132,1156,1.688336,1.688336
1,53243585a1d6dc2643021fd1853d8905,222776.05,358,410,1.639066,3.327402
2,4a3ca9315b744ce9f8e9374361493884,200472.92,1806,1987,1.474972,4.802374
3,fa1c13f2614d7b5c4749cbc52fecda94,194042.03,585,586,1.427657,6.230031
4,7c67e1448b00f6e969d365cea6b010ab,187923.89,982,1364,1.382643,7.612674
...,...,...,...,...,...,...
3090,34aefe746cd81b7f3b23253ea28bef39,8.00,1,1,0.000059,99.999820
3091,702835e4b785b67a084280efca355756,7.60,1,1,0.000056,99.999876
3092,1fa2d3def6adfa70e58c276bb64fe5bb,6.90,1,1,0.000051,99.999926
3093,77128dec4bec4878c37ab7d6169d6f26,6.50,1,1,0.000048,99.999974


In [147]:
# Top 1%
top_1 = int(len(seller_risk) * 0.01)
top_1_pct_revenue = (
    seller_risk.head(top_1)['revenue_pct'].sum()
)

# Top 5%
top_5 = int(len(seller_risk) * 0.05)
top_5_pct_revenue = (
    seller_risk.head(top_5)['revenue_pct'].sum()
)

# Top 10%
top_10 = int(len(seller_risk) * 0.10)
top_10_pct_revenue = (
    seller_risk.head(top_10)['revenue_pct'].sum()
)

# Sellers generating 80% revenue
sellers_for_80_pct = len(
    seller_risk[
        seller_risk['cumulative_pct'] <= 80
    ]
)

print("Top 1% Revenue:", round(top_1_pct_revenue,2))
print("Top 5% Revenue:", round(top_5_pct_revenue,2))
print("Top 10% Revenue:", round(top_10_pct_revenue,2))
print("Sellers for 80% Revenue:", sellers_for_80_pct)

Top 1% Revenue: 25.67
Top 5% Revenue: 53.18
Top 10% Revenue: 67.49
Sellers for 80% Revenue: 543


# **Marketplace Revenue Concentration by Seller Rank**

In [148]:
import plotly.express as px

fig = px.line(
    seller_risk,
    x=seller_risk.index + 1,
    y='cumulative_pct',
    title='Marketplace Revenue Concentration by Seller Rank'
)

# 80% reference line
fig.add_hline(
    y=80,
    line_dash='dash',
    line_color='red',
    annotation_text='80% Revenue'
)

# 543 sellers reference line
fig.add_vline(
    x=543,
    line_dash='dash',
    line_color='green',
    annotation_text='543 Sellers'
)

fig.update_traces(
    line_width=3
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis_title='Seller Rank',
    yaxis_title='Cumulative Revenue (%)',
    xaxis=dict(
        showgrid=False,
        showline=True,
        linecolor='black'
    ),
    yaxis=dict(
        showgrid=True,
        gridcolor='#E5E5E5',
        showline=True,
        linecolor='black',
        range=[0, 100]
    )
)

fig.show()

Marketplace revenue is moderately concentrated across the seller ecosystem. Approximately 543 sellers (17.5% of all sellers) generate 80% of total marketplace revenue, while the top 10% of sellers contribute nearly 67.5% of revenue. Although a relatively small group of sellers drives marketplace performance, revenue generation remains distributed across hundreds of active sellers, reducing dependency on a handful of dominant vendors.

# **Seller Category Specialization**

In [150]:
seller_category = (
    product_sales
    .groupby(
        ['product_category_name_english', 'seller_id']
    )
    .agg(
        revenue=('price', 'sum')
    )
    .reset_index()
)

seller_category.head()

,product_category_name_english,seller_id,revenue
0,agro_industry_and_commerce,06579cb253ecd5a3a12a9e6eb6bf8f47,186.80
1,agro_industry_and_commerce,0ed6ce5d87fd9c69eaacaeb778d67235,544.60
2,agro_industry_and_commerce,1996942dc085d7773ba77a529b163cd0,559.99
3,agro_industry_and_commerce,2528744c5ef5d955adc318720a94d2e7,6245.00
4,agro_industry_and_commerce,269cff2d3c8d205c11f37a52402ea93b,721.69


In [152]:
category_specialization = (
    seller_category
    .groupby('product_category_name_english')
    .agg(
        total_revenue=('revenue', 'sum'),
        top_seller_revenue=('revenue', 'max'),
        seller_count=('seller_id', 'nunique')
    )
    .reset_index()
)

In [153]:
category_specialization['top_seller_share'] = (
    category_specialization['top_seller_revenue']
    / category_specialization['total_revenue']
) * 100

In [154]:
category_specialization.sort_values(
    'top_seller_share',
    ascending=False
).head(15)

,product_category_name_english,total_revenue,top_seller_revenue,seller_count,top_seller_share
11,cds_dvds_musicals,730.00,730.00,1,100.000000
52,la_cuisine,2054.99,1935.00,2,94.161042
14,computers,222963.13,166871.75,9,74.842755
67,tablets_printing_image,7528.41,4993.80,6,66.332732
61,security_and_services,283.29,183.29,2,64.700484
57,office_furniture,273960.70,172959.66,34,63.133019
64,small_appliances_home_oven_and_coffee,47445.71,25222.60,15,53.160971
35,flowers,1110.04,569.80,3,51.331484
58,party_supplies,4485.18,2239.40,12,49.928877
25,dvds_blu_ray,5999.39,2840.00,10,47.338146


In [159]:
category_specialization_filtered = category_specialization[
    (category_specialization['seller_count'] >= 10) &
    (category_specialization['total_revenue'] >= 10000)]

In [175]:
category_specialization_filtered.sort_values(
    'top_seller_share',
    ascending=False
).head(10)

,product_category_name_english,total_revenue,top_seller_revenue,seller_count,top_seller_share
39,furniture_decor,729762.49,111530.45,370,15.283116
28,fashion_bags_accessories,152823.54,18076.83,99,11.828564
54,market_place,28378.47,3328.70,59,11.729667
5,auto,592720.11,64925.30,383,10.953787
26,electronics,160246.74,14178.09,149,8.847662
60,pet_shop,214315.41,18263.74,137,8.521898
43,health_beauty,1258681.34,79284.55,492,6.299017
15,computers_accessories,911954.32,53257.52,287,5.839933
65,sports_leisure,988048.97,54056.28,481,5.471012
49,housewares,632248.66,30844.00,468,4.878460


In [176]:
top_share_chart = (
    category_specialization_filtered
    .sort_values(
        'top_seller_share',
        ascending=False
    )
    .head(15)
)

In [177]:
fig = px.bar(
    top_share_chart,
    x='top_seller_share',
    y='product_category_name_english',
    orientation='h',
    title='Category Revenue Dominance by Top Seller'
)

fig.update_yaxes(autorange='reversed')

fig.show()

# **Delivery & Customer Experience Analytics**

In [179]:
orders.shape


(96478, 8)

In [180]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:00,02-10-2017 11:07,04-10-2017 19:55,10-10-2017 21:25,18-10-2017 00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:00,26-07-2018 03:24,26-07-2018 14:31,07-08-2018 15:27,13-08-2018 00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:00,08-08-2018 08:55,08-08-2018 13:50,17-08-2018 18:06,04-09-2018 00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:00,18-11-2017 19:45,22-11-2017 13:39,02-12-2017 00:28,15-12-2017 00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:00,13-02-2018 22:20,14-02-2018 19:46,16-02-2018 18:17,26-02-2018 00:00


In [183]:
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in date_cols:
    orders[col] = pd.to_datetime(
        orders[col],
        dayfirst=True,
        errors='coerce'
    )

In [184]:
orders[date_cols].dtypes

,0
order_purchase_timestamp,datetime64[ns]
order_approved_at,datetime64[ns]
order_delivered_carrier_date,datetime64[ns]
order_delivered_customer_date,datetime64[ns]
order_estimated_delivery_date,datetime64[ns]


In [185]:
orders['approval_days'] = (
    orders['order_approved_at']
    - orders['order_purchase_timestamp']
).dt.total_seconds() / 86400

orders['dispatch_days'] = (
    orders['order_delivered_carrier_date']
    - orders['order_approved_at']
).dt.total_seconds() / 86400

orders['delivery_days'] = (
    orders['order_delivered_customer_date']
    - orders['order_delivered_carrier_date']
).dt.total_seconds() / 86400

orders['cycle_time_days'] = (
    orders['order_delivered_customer_date']
    - orders['order_purchase_timestamp']
).dt.total_seconds() / 86400

In [186]:
velocity_summary = pd.DataFrame({
    'Metric': [
        'Approval Time',
        'Dispatch Time',
        'Delivery Time',
        'Total Cycle Time'
    ],
    'Average Days': [
        orders['approval_days'].mean(),
        orders['dispatch_days'].mean(),
        orders['delivery_days'].mean(),
        orders['cycle_time_days'].mean()
    ]
})

velocity_summary.round(2)

,Metric,Average Days
0,Approval Time,0.43
1,Dispatch Time,2.80
2,Delivery Time,9.33
3,Total Cycle Time,12.56


In [187]:
import plotly.express as px

velocity_summary = pd.DataFrame({
    'Stage': [
        'Approval',
        'Dispatch',
        'Delivery'
    ],
    'Days': [
        0.43,
        2.80,
        9.33
    ]
})

fig = px.bar(
    velocity_summary,
    x='Stage',
    y='Days',
    title='Average Time Spent Across Fulfillment Stages',
    text='Days'
)

fig.update_traces(
    texttemplate='%{text:.2f}',
    textposition='outside'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)

fig.show()

The average order cycle time was approximately 12.6 days, with delivery transit time accounting for nearly 74% of the fulfillment process. While order approval (0.43 days) and dispatch activities (2.8 days) were relatively efficient, the majority of delays occurred during the final delivery stage, indicating that transportation and last-mile logistics represent the largest opportunity for operational improvement.

# **Delivery Reliability Analysis**

In [188]:
#Calculate Delay Days
orders['delay_days'] = (
    orders['order_delivered_customer_date']
    - orders['order_estimated_delivery_date']
).dt.total_seconds() / 86400

In [189]:
#Create Delivery Status
orders['delivery_status'] = np.where(
    orders['delay_days'] > 0,
    'Late',
    'On Time / Early'
)

In [190]:
#Reliability Summary
reliability_summary = (
    orders['delivery_status']
    .value_counts(normalize=True)
    .mul(100)
    .reset_index()
)

reliability_summary.columns = [
    'Delivery Status',
    'Percentage'
]

reliability_summary

,Delivery Status,Percentage
0,On Time / Early,91.888306
1,Late,8.111694


In [191]:
#Average Delay
orders['delay_days'].describe()

,delay_days
count,96470.000000
mean,-11.178511
std,10.184352
min,-146.016667
25%,-16.244444
50%,-11.948611
75%,-6.390278
max,188.975000


In [192]:
#Late Deliveries Only
late_orders = orders[
    orders['delay_days'] > 0
]

late_orders['delay_days'].describe()

,delay_days
count,7826.000000
mean,9.551386
std,13.952540
min,0.002083
25%,1.866667
50%,5.806250
75%,11.820660
max,188.975000


In [193]:
fig = px.pie(
    reliability_summary,
    names='Delivery Status',
    values='Percentage',
    title='On-Time vs Late Deliveries'
)

fig.update_layout(
    title_x=0.5
)

fig.show()

In [195]:
late_orders['delay_days'].describe().round(2)

,delay_days
count,7826.00
mean,9.55
std,13.95
min,0.00
25%,1.87
50%,5.81
75%,11.82
max,188.98


Delivery performance was highly reliable, with nearly 92% of orders delivered on or before the promised date. However, the small proportion of delayed orders experienced significant disruptions, with an average delay of approximately 10 days and some orders delayed by more than 180 days. This suggests that while delivery failures were relatively uncommon, their impact on affected customers was potentially severe.

# **Delivery Performance vs Customer Reviews**

In [197]:
reviews = pd.read_csv(
    '/content/drive/MyDrive/Ecommerce analysis/olist_order_reviews_dataset.csv'
)

reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,18-01-2018 00:00,18-01-2018 21:46
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,10-03-2018 00:00,11-03-2018 03:05
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,17-02-2018 00:00,18-02-2018 14:36
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,21-04-2017 00:00,21-04-2017 22:02
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,01-03-2018 00:00,02-03-2018 10:26


In [198]:
reviews.columns.tolist()

['review_id',
 'order_id',
 'review_score',
 'review_comment_title',
 'review_comment_message',
 'review_creation_date',
 'review_answer_timestamp']

In [199]:
#Merge Orders + Reviews
delivery_reviews = orders.merge(
    reviews[['order_id', 'review_score']],
    on='order_id',
    how='inner'
)

delivery_reviews.shape

(96361, 15)

In [200]:
#Create Delivery Status


delivery_reviews['delivery_status'] = np.where(
    delivery_reviews['delay_days'] > 0,
    'Late',
    'On Time / Early'
)

In [201]:
#Compare Review Scores
review_impact = (
    delivery_reviews
    .groupby('delivery_status')
    .agg(
        avg_review=('review_score', 'mean'),
        orders=('order_id', 'count')
    )
    .reset_index()
)

review_impact

,delivery_status,avg_review,orders
0,Late,2.566494,7700
1,On Time / Early,4.293737,88661


In [202]:
#severity groups

conditions = [
    delivery_reviews['delay_days'] <= 0,
    delivery_reviews['delay_days'].between(0, 5),
    delivery_reviews['delay_days'].between(5, 15),
    delivery_reviews['delay_days'] > 15
]

labels = [
    'On Time / Early',
    '1-5 Days Late',
    '5-15 Days Late',
    '15+ Days Late'
]

delivery_reviews['delay_group'] = np.select(
    conditions,
    labels,
    default='Unknown'
)

In [203]:
#Review Score by Delay Severity
delay_reviews = (
    delivery_reviews
    .groupby('delay_group')
    .agg(
        avg_review=('review_score', 'mean'),
        orders=('order_id', 'count')
    )
    .reset_index()
)

delay_reviews

,delay_group,avg_review,orders
0,1-5 Days Late,3.460658,3584
1,15+ Days Late,1.724907,1345
2,5-15 Days Late,1.818477,2771
3,On Time / Early,4.293718,88653
4,Unknown,4.500000,8


In [204]:
fig = px.bar(
    delay_reviews,
    x='delay_group',
    y='avg_review',
    text='avg_review',
    title='Customer Review Score by Delivery Delay Severity'
)

fig.update_traces(
    texttemplate='%{text:.2f}',
    textposition='outside'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)

fig.show()

Late deliveries are not merely operational failures; they have a direct and measurable impact on customer satisfaction. Orders delivered on time received an average review score of 4.29, while delayed orders averaged only 2.57. Customer ratings declined progressively as delay severity increased, falling to approximately 1.7 stars for orders delayed by more than 15 days.

# **Revenue Trend Analysis**

In [205]:
revenue_orders = payments.merge(
    orders[
        [
            'order_id',
            'order_purchase_timestamp'
        ]
    ],
    on='order_id',
    how='inner'
)

revenue_orders.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value,order_purchase_timestamp
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33,2018-04-25 22:01:00
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39,2018-06-26 11:01:00
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71,2017-12-12 11:19:00
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78,2017-12-06 12:04:00
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45,2018-05-21 13:59:00


In [206]:
#Create Revenue Month
revenue_orders['year_month'] = (
    revenue_orders['order_purchase_timestamp']
    .dt.to_period('M')
    .astype(str)
)

In [207]:
#Monthly Revenue
monthly_revenue = (
    revenue_orders
    .groupby('year_month')
    .agg(
        revenue=('payment_value', 'sum')
    )
    .reset_index()
)

monthly_revenue.head()

,year_month,revenue
0,2016-10,46566.71
1,2016-12,19.62
2,2017-01,127545.67
3,2017-02,271298.65
4,2017-03,414369.39


In [208]:
#Growth Rate
monthly_revenue['growth_pct'] = (
    monthly_revenue['revenue']
    .pct_change()
    * 100
)

In [209]:
monthly_revenue

,year_month,revenue,growth_pct
0,2016-10,46566.71,NaN
1,2016-12,19.62,-99.957867
2,2017-01,127545.67,649979.867482
3,2017-02,271298.65,112.707064
4,2017-03,414369.39,52.735515
5,2017-04,390952.18,-5.651289
6,2017-05,567066.73,45.047594
7,2017-06,490225.60,-13.550633
8,2017-07,566403.93,15.539443
9,2017-08,646000.61,14.052989


In [210]:
monthly_revenue.head()

,year_month,revenue,growth_pct
0,2016-10,46566.71,NaN
1,2016-12,19.62,-99.957867
2,2017-01,127545.67,649979.867482
3,2017-02,271298.65,112.707064
4,2017-03,414369.39,52.735515


In [211]:
monthly_revenue.tail()

,year_month,revenue,growth_pct
17,2018-04,1132933.95,1.093619
18,2018-05,1128836.69,-0.361650
19,2018-06,1012090.68,-10.342152
20,2018-07,1027903.86,1.562427
21,2018-08,985414.28,-4.133614


In [213]:
monthly_revenue = monthly_revenue[
    monthly_revenue['year_month'] >= '2017-01'
].copy()

monthly_revenue['growth_pct'] = (
    monthly_revenue['revenue']
    .pct_change()
    * 100
)

In [217]:
monthly_revenue['rolling_3m'] = (
    monthly_revenue['revenue']
    .rolling(3)
    .mean()
)

In [218]:
fig = px.line(
    monthly_revenue,
    x='year_month',
    y='revenue',
    markers=True,
    title='Monthly Marketplace Revenue Trend'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)
y=['revenue', 'rolling_3m']

fig.show()

Marketplace revenue demonstrated rapid growth throughout 2017, increasing nearly sixfold before reaching a peak of approximately ₹1.15M in November 2017. Following a seasonal spike, revenue stabilized above ₹1M per month during 2018, indicating a transition from an expansion phase to a more mature and consistent revenue-generating marketplace.

# **Revenue Concentration Analysis**

In [220]:
#Customer Revenue Table
customer_revenue = (
    orders[['order_id', 'customer_id']]
    .merge(
        payments[['order_id', 'payment_value']],
        on='order_id'
    )
)

In [221]:
customer_revenue = (
    customer_revenue
    .groupby('customer_id')
    .agg(
        revenue=('payment_value', 'sum')
    )
    .reset_index()
)

In [222]:
#Pareto Logic
customer_revenue = (
    customer_revenue
    .sort_values(
        'revenue',
        ascending=False
    )
)

In [223]:
customer_revenue['revenue_pct'] = (
    customer_revenue['revenue']
    /
    customer_revenue['revenue'].sum()
) * 100

In [224]:
customer_revenue['cumulative_pct'] = (
    customer_revenue['revenue_pct']
    .cumsum()
)

In [228]:
#Customers Needed for 80%
customers_for_80_pct = len(
    customer_revenue[
        customer_revenue['cumulative_pct'] <= 80
    ]
)



In [227]:
len(customer_revenue)

96477

In [229]:
#Final Revenue Concentration Summary


concentration_summary = pd.DataFrame({
    'Entity': ['Categories', 'Sellers', 'Customers'],
    'Total_Count': [71, 3095, 96477],
    'Count_For_80pct_Revenue': [16, 543, 47480]
})

concentration_summary['Pct_For_80pct_Revenue'] = (
    concentration_summary['Count_For_80pct_Revenue']
    / concentration_summary['Total_Count']
    * 100
).round(1)

concentration_summary

,Entity,Total_Count,Count_For_80pct_Revenue,Pct_For_80pct_Revenue
0,Categories,71,16,22.5
1,Sellers,3095,543,17.5
2,Customers,96477,47480,49.2


In [232]:
fig = px.bar(
    concentration_summary,
    x='Entity',
    y='Pct_For_80pct_Revenue',
    text='Pct_For_80pct_Revenue',
    color='Pct_For_80pct_Revenue',
    color_continuous_scale='Blues',
    title='Percentage of Entities Required to Generate 80% of Revenue'
)

fig.update_traces(
    texttemplate='%{text:.1f}%',
    textposition='outside'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white',
    coloraxis_showscale=False,
    yaxis_title='% Required for 80% Revenue'
)

fig.show()

Marketplace revenue is concentrated around a limited set of product categories and high-performing sellers, but customer spending remains broadly distributed across the customer base. While operational dependence exists on key categories and sellers, the marketplace exhibits low customer concentration risk, indicating a diversified and resilient demand base.

# **Payment Behavior Analysis**

In [233]:
payments.columns.tolist()

['order_id',
 'payment_sequential',
 'payment_type',
 'payment_installments',
 'payment_value']

In [235]:
#Payment Method Analysis


payment_method = (
    payments
    .groupby('payment_type')
    .agg(
        orders=('order_id','count'),
        revenue=('payment_value','sum')
    )
    .reset_index()
)

#Revenue Share:

payment_method['revenue_pct'] = (
    payment_method['revenue']
    /
    payment_method['revenue'].sum()
    * 100
)

#Sort:

payment_method = payment_method.sort_values(
    'revenue',
    ascending=False
)

payment_method

,payment_type,orders,revenue,revenue_pct
1,credit_card,76795,12542084.19,78.344584
0,boleto,19784,2869361.27,17.923569
4,voucher,5775,379436.87,2.370166
2,debit_card,1529,217989.79,1.361681
3,not_defined,3,0.00,0.000000


In [236]:
fig = px.bar(
    payment_method,
    x='payment_type',
    y='revenue_pct',
    text='revenue_pct',
    title='Revenue Share by Payment Method'
)

fig.update_traces(
    texttemplate='%{text:.1f}%',
    textposition='outside'
)

fig.update_layout(
    title_x=0.5,
    plot_bgcolor='white',
    paper_bgcolor='white'
)

fig.show()

Credit cards dominated marketplace transactions, accounting for approximately 78.3% of total revenue, highlighting strong customer preference for card-based payments. Boleto emerged as the primary alternative payment method, contributing nearly 18% of revenue, while debit cards and vouchers represented only a small share of marketplace transactions.